tensordict/nn/common.py

```diff
@@ -771,29 +771,47 @@ def reset_out_keys(self):
                 del self._forward_hooks[i]
         return self
 
-    def reset_parameters(self):
+    def reset_parameters(self, parameters: Optional[TensorDictBase] = None) -> None:
         """Recursively reset the parameters of the module and its children.
 
+        Args:
+            parameters: If set to None, the module will reset using self.parameters().
+                Otherwise, we will reset the parameters in the tensordict in-place. This is
+                useful for functional modules where the parameters are not stored in the module itself.
+
         Examples:
         >>> from tensordict.nn import TensorDictModule
         >>> from torch import nn
         >>> net = nn.Sequential(nn.Linear(2,3), nn.ReLU())
         >>> old_param = net[0].weight.clone()
         >>> module = TensorDictModule(net, in_keys=['bork'], out_keys=['dork'])
         >>> module.reset_parameters()
-        >>> (old_param == net[0].weight).all()
+        >>> (old_param == net[0].weight).any()
         tensor(False)
+
+        >>> from tensordict import TensorDict
+        >>> from tensordict.nn import TensorDictModule
+        >>> from torch import nn
+        >>> net = nn.Sequential(nn.Linear(2,3), nn.ReLU())
+        >>> module = TensorDictModule(net, in_keys=['bork'], out_keys=['dork'])
+        >>> params = TensorDict.from_module(module)
+        >>> old_params = params.clone(recurse=True)
+        >>> module.reset_parameters(params)
+        >>> (old_params == params).any()
+        False
         """
-        for m in self.children():
-            self._reset_parameters(m)
+        if parameters is not None:
+            sanitized_parameters = parameters.apply(
+                lambda x: x.detach().requires_grad_(), inplace=False
+            )
+            repopulate_module(self, sanitized_parameters)
 
-    def _reset_parameters(self, module: Union[nn.Module, TensorDictModuleBase]):
-        if isinstance(module, nn.Module):
-            if hasattr(module, "reset_parameters"):
-                module.reset_parameters()
-            else:
-                for m in module.children():
-                    self._reset_parameters(m)
+        for child in list(self.children()):
+            child.apply(
+                lambda m: m.reset_parameters()
+                if hasattr(m, "reset_parameters")
+                else None
+            )
```

Comment:
Just noticed, why not

```suggestion
        def _reset(module):
            if hasattr(module, 'reset_parameters'):
                module.reset_parameters()
        self.apply(reset_parameters)
```